Makefile

```diff
@@ -59,8 +69,17 @@ image-cephcsi: cephcsi
 	$(CONTAINER_CMD) build -t $(CSI_IMAGE_NAME):$(CSI_IMAGE_VERSION) deploy/cephcsi/image
 
 push-image-cephcsi: image-cephcsi
-	$(CONTAINER_CMD) push $(CSI_IMAGE_NAME):$(CSI_IMAGE_VERSION)
-
+	$(CONTAINER_CMD) tag $(CSI_IMAGE_NAME):$(CSI_IMAGE_VERSION) $(CSI_IMAGE_NAME)-$(ARCH):$(CSI_IMAGE_VERSION)
+	$(CONTAINER_CMD) push $(CSI_IMAGE_NAME)-$(ARCH):$(CSI_IMAGE_VERSION)
+
+# "docker manifest" requires experimental feature enabled
+push-manifest-cephcsi: export DOCKER_CLI_EXPERIMENTAL=enabled
+push-manifest-cephcsi:
+	$(CONTAINER_CMD) manifest create \
```

Comment:
Can this be part of `push-image-cephcsi` which will be called when the PR is merged?

Reply:
The problem is that push-manifest-cephcsi should run after both x86 and arm successfully run push-image-cephcsi.

push-image-cephcsi are launched in parallel on x86 and arm nodes to build arch dependent image separately. We don't know which one finishes early.
After they all succeed, push-manifest-cephcsi job is started on one x86 node to update manifest to refer to new built images.

Reply:
i have done some testing below are for the reference

https://github.com/Madhu-1/test-travis
https://hub.docker.com/r/madhupr001/multiarch/tags

```
docker manifest inspect madhupr001/multiarch

{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1374,
         "digest": "sha256:fea065be9e2157d01ebd157130fad27cd977e50044c77ad8c2be456733de109a",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1374,
         "digest": "sha256:fea065be9e2157d01ebd157130fad27cd977e50044c77ad8c2be456733de109a",
         "platform": {
            "architecture": "arm",
            "os": "linux",
            "variant": "v7"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 1374,
         "digest": "sha256:fea065be9e2157d01ebd157130fad27cd977e50044c77ad8c2be456733de109a",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}
```

Reply:
Cool! I don't know we can easily cross build go program with GOARCH. It can simplify the code much. No new job is necessary on Arm node.
Thanks @Madhu-1, will update per your reference.

Reply:
Tested this approach and found some issues.

GOARCH helps to cross build Arm cephcsi image on x86 easily. But [Dockerfile](https://github.com/ceph/ceph-csi/blob/master/deploy/cephcsi/image/Dockerfile) is not easy to handle.
```
FROM ceph/ceph:v14.2

ENV CSIBIN=/usr/local/bin/cephcsi

COPY cephcsi $CSIBIN

RUN chmod +x $CSIBIN

ENTRYPOINT ["/usr/local/bin/cephcsi"]
```
When running on x86, FROM always pulls x86 base image. So the final Arm image contains x86 ceph/ceph base image and Arm cephcsi binary, which is problematic. Image seems okay as entrypoint is Arm binary. Try starting the image on Arm host with "--entrypoint /bin/bash" leads to "exec format error".

To fix this issues, we can replace base image on the fly before building Arm image. Rook uses similar trick. See [1](https://github.com/rook/rook/blob/master/images/ceph/Dockerfile#L16), [2](https://github.com/rook/rook/blob/master/images/ceph/Makefile#L52).

After fixing it, I met with another issue at "RUN chmod +x $CSIBIN". The "chown" binary is from ceph/ceph base image, so it's now an Arm binary and cannot run x86 host. Though it's possible to workaround, things may break if we add new commands later.

@Madhu-1, comments?

Reply:
we can follow the similar approach what rook follows,
we need to add a Test  in Travis to use new arm image in ARM architecture

Reply:
To test cephcsi Arm image is not easy now.
Current test script uses minikube to launch a k8s storage cluster for testing purpose.
It requires some K8s csi sidecar images which are missing on Arm, such as quay.io/k8scsi/csi-provisioner, quay.io/k8scsi/csi-snapshotter, quay.io/k8scsi/csi-attacher. Without them, cephcsi cannot be validated.

Is it acceptable to only build cephcsi Arm image without testing for now(or add a simpler validation for Arm)? If not, I think we have to enable Arm [k8scsi](https://github.com/kubernetes-csi) images first.

Reply:
okay, I think we can add the code for testing it and comment it as of now and enable it later when we have the container for ARM architecture?

Reply:
Yes, will try.